src/algebra/pointwise.lean

```diff
@@ -838,7 +840,113 @@ s.mul_zero_subset.antisymm $ by simpa [finset.mem_mul] using hs
+by simp [subset_iff, mem_mul]
+
 end mul_zero_class
+end decidable_eq
+
+open_locale pointwise
+variables {u : finset α} {b : α} {x y : β}
+
+@[to_additive]
+lemma singleton_one [has_one α] : ({1} : finset α) = 1 := rfl
+
+@[to_additive]
+lemma one_mem_one [has_one α] : (1 : α) ∈ (1 : finset α) := by simp [has_one.one]
+
+@[to_additive]
+theorem one_nonempty [has_one α] : (1 : finset α).nonempty := ⟨1, one_mem_one⟩
+
+@[simp, to_additive]
+theorem image_one [decidable_eq β] [has_one α] {f : α → β} : image f 1 = {f 1} :=
+image_singleton f 1
+
+@[to_additive add_image_prod]
+lemma image_mul_prod [decidable_eq α] [has_mul α] :
+  image (λ x : α × α, x.fst * x.snd) (s.product t) = s * t := rfl
+
+@[simp, to_additive]
+lemma image_mul_left [decidable_eq α] [group α] :
+  image (λ b, a * b) t = preimage t (λ b, a⁻¹ * b) (assume x hx y hy, (mul_right_inj a⁻¹).mp) :=
+coe_injective $ by simp
+
+@[simp, to_additive]
+lemma image_mul_right [decidable_eq α] [group α] :
+  image (λ a, a * b) t = preimage t (λ a, a * b⁻¹) (assume x hx y hy, (mul_left_inj b⁻¹).mp) :=
+coe_injective $ by simp
+
+@[to_additive]
+lemma image_mul_left' [decidable_eq α] [group α] :
+  image (λ b, a⁻¹ * b) t = preimage t (λ b, a * b) (assume x hx y hy, (mul_right_inj a).mp) :=
+by simp
+
+@[to_additive]
+lemma image_mul_right' [decidable_eq α] [group α] :
+  image (λ a, a * b⁻¹) t = preimage t (λ a, a * b) (assume x hx y hy, (mul_left_inj b).mp) :=
+by simp
+
+@[simp, to_additive]
+lemma preimage_mul_left_singleton [group α] :
+  preimage {b} ((*) a) (assume x hx y hy, (mul_right_inj a).mp) = {a⁻¹ * b} :=
+by { classical, rw [← image_mul_left', image_singleton] }
+
+@[simp, to_additive]
+lemma preimage_mul_right_singleton [group α] :
+  preimage {b} (* a) (assume x hx y hy, (mul_left_inj a).mp) = {b * a⁻¹} :=
+by { classical, rw [← image_mul_right', image_singleton] }
+
+@[simp, to_additive]
+lemma preimage_mul_left_one [group α] :
+  preimage 1 (λ b, a * b) (assume x hx y hy, (mul_right_inj a).mp) = {a⁻¹} :=
+by {classical, rw [← image_mul_left', image_one, mul_one] }
+
+@[simp, to_additive]
+lemma preimage_mul_right_one [group α] :
+  preimage 1 (λ a, a * b) (assume x hx y hy, (mul_left_inj b).mp) = {b⁻¹} :=
+by {classical, rw [← image_mul_right', image_one, one_mul] }
+
+@[to_additive]
+lemma preimage_mul_left_one' [group α] :
+  preimage 1 (λ b, a⁻¹ * b) (assume x hx y hy, (mul_right_inj _).mp) = {a} := by simp
+
+@[to_additive]
+lemma preimage_mul_right_one' [group α] :
+  preimage 1 (λ a, a * b⁻¹) (assume x hx y hy, (mul_left_inj _).mp) = {b} := by simp
+
+@[to_additive]
+protected lemma mul_comm [decidable_eq α] [comm_semigroup α] : s * t = t * s :=
+by exact_mod_cast @set.mul_comm _ (s : set α) t _
+
+/-- `finset α` is a `mul_one_class` under pointwise operations if `α` is. -/
+@[to_additive /-"`finset α` is an `add_zero_class` under pointwise operations if `α` is."-/]
+protected def mul_one_class [decidable_eq α] [mul_one_class α] : mul_one_class (finset α) :=
```

Comment:
Could you please use `function.injective.*` constructors here and below?

Reply:
oh nice, thanks!